### PR TITLE
Graph directionality

### PR DESCRIPTION
**Pull request by MalteMederacke**
Function to convert the input graph into a directed graph and update the skeleton object

### CODE REVIEW

File: src/skeleplex/graph/skeleton_graph.py [hunk 1]
```diff
@@ -50,6 +50,52 @@ def skeleton_graph_decoder(json_object):
...
+        fragments = None
+
+    di_graph = nx.DiGraph(graph)
+    di_graph.remove_edges_from(di_graph.edges - nx.bfs_edges(di_graph, origin))
+
+    if fragments:
+        # choose a node with the highest degree as the origin node
+        # This could lead to the wrong edges being removed, but it is a good start
+        for fragment in nx.connected_components(fragments):
+            fragment_subgraph = fragments.subgraph(fragment)
```
Comment (kevinyamauchi) [inline src/skeleplex/graph/skeleton_graph.py:94]: Does networkx have a way to directly iterate over connected components? It might make this a bit cleaner if so. No worries if not.

File: src/skeleplex/graph/skeleton_graph.py [hunk 2]
```diff
@@ -50,6 +50,52 @@ def skeleton_graph_decoder(json_object):
...
+            origin = max(fragment_subgraph.degree, key=lambda x: x[1])[0]
```
Comment (kevinyamauchi) [inline src/skeleplex/graph/skeleton_graph.py:98]: can you please add a comment explaining this line?

File: tests/graph/test_skeleton_graph.py [hunk 3]
```diff
@@ -31,3 +31,11 @@ def test_skeleton_graph_json_round_trip(simple_t_skeleton_graph, tmp_path):
 
     # check that the graphs are equal
     assert simple_t_skeleton_graph == new_skeleton_graph
+
+def test_skeleton_graph_to_directed(simple_t_skeleton_graph):
+    """Test converting a SkeletonGraph to a directed graph."""
+    directed_graph = simple_t_skeleton_graph.graph.to_directed()
```
Comment (kevinyamauchi) [inline tests/graph/test_skeleton_graph.py]: Does this work? It looks like `origin` is a required argument

File: tests/graph/test_skeleton_graph.py [hunk 4]
```diff
@@ -31,3 +31,11 @@ def test_skeleton_graph_json_round_trip(simple_t_skeleton_graph, tmp_path):
...
+    assert directed_graph.is_directed()
+
+    # check that the directed graph has the same nodes
+    assert set(directed_graph.nodes) == set(simple_t_skeleton_graph.graph.nodes)
```
Comment (kevinyamauchi) [inline tests/graph/test_skeleton_graph.py:42]: can you please also add a test that shows that the edge directionality is correct?

File: src/skeleplex/graph/skeleton_graph.py [hunk 5]
```diff
@@ -50,6 +50,52 @@ def skeleton_graph_decoder(json_object):
     return json_object
 
 
+def make_graph_directed(graph: nx.Graph, origin=int) -> nx.DiGraph:
+    """Return a directed graph from an undirected graph.
+
+    The directed graph has the same nodes and edges as the undirected graph.
```
Comment (kevinyamauchi) [inline src/skeleplex/graph/skeleton_graph.py:60]: Can you please add an explanation of the behavior when there are "fragments"?

**Review by kevinyamauchi — COMMENTED** [review]
I think this looks good. I left a few minor comments - mostly about documentation. I think we can merge once you address those. Thanks!

File: src/skeleplex/graph/skeleton_graph.py [hunk 6]
```diff
@@ -50,6 +50,52 @@ def skeleton_graph_decoder(json_object):
...
+        The undirected graph to convert to a directed graph.
+    origin : int
+        The node to use as the origin node for the directed graph.
+        The origin node will have no incoming edges.
+    """
+    if isinstance(graph, nx.DiGraph):
+        print("The input graph is already a directed graph.")
+        return graph
+    if len(list(nx.connected_components(graph))) > 1:
+        Warning("""
```
Comment (kevinyamauchi) [inline src/skeleplex/graph/skeleton_graph.py]: Can you please use warnings.warn instead? I don't think the `Warning` class will actually emit the warning by itself.

https://docs.python.org/3/library/warnings.html#warnings.warn

File: src/skeleplex/graph/skeleton_graph.py [hunk 7]
```diff
@@ -50,6 +50,52 @@ def skeleton_graph_decoder(json_object):
...
+    Parameters
+    ----------
+    graph : nx.Graph
+        The undirected graph to convert to a directed graph.
+    origin : int
+        The node to use as the origin node for the directed graph.
+        The origin node will have no incoming edges.
+    """
+    if isinstance(graph, nx.DiGraph):
+        print("The input graph is already a directed graph.")
```
Comment (kevinyamauchi) [inline src/skeleplex/graph/skeleton_graph.py]: Can you please use the logging module? That way users can filter this message out (e.g., when they are running a pipeline). The "info" level is probably fine.

https://docs.python.org/3/library/logging.html

File: src/skeleplex/graph/skeleton_graph.py [hunk 8]
```diff
@@ -50,6 +50,52 @@ def skeleton_graph_decoder(json_object):
     return json_object
 
 
+def make_graph_directed(graph: nx.Graph, origin=int) -> nx.DiGraph:
```
Comment (kevinyamauchi) [inline src/skeleplex/graph/skeleton_graph.py]:
```suggestion
def make_graph_directed(graph: nx.Graph, origin: int) -> nx.DiGraph:
```

File: src/skeleplex/graph/skeleton_graph.py [hunk 9]
```diff
@@ -156,3 +202,17 @@ def __eq__(self, other: "SkeletonGraph"):
             return False
         else:
             return True
+
+    def to_directed(self, origin=int) -> nx.DiGraph:
```
Comment (kevinyamauchi) [inline src/skeleplex/graph/skeleton_graph.py]:
```suggestion
    def to_directed(self, origin: int) -> nx.DiGraph:
```

File: tests/graph/test_skeleton_graph.py (hunk 3 above)
Comment (kevinyamauchi) [reply tests/graph/test_skeleton_graph.py]: Ah I see. I think it was running because the `origin` parameter was assigned `int` instead of typed as `int` (i.e., `=` instead of `:`).

Comment (MalteMederacke) [reply tests/graph/test_skeleton_graph.py]: woops, that was because I called the networkx function on the graph and not on the skeleton object. Good catch

File: src/skeleplex/graph/skeleton_graph.py (hunk 1 above)
Comment (MalteMederacke) [reply src/skeleplex/graph/skeleton_graph.py:94]: https://networkx.org/documentation/networkx-1.9/reference/generated/networkx.algorithms.components.connected.connected_component_subgraphs.html

has been depreciated, I could do list comprehension, if necessary

File: src/skeleplex/graph/skeleton_graph.py (hunk 2 above)
Comment (MalteMederacke) [reply src/skeleplex/graph/skeleton_graph.py:98]: Done, this approach is for sure not optimal, but I don't see an easy solution. Alternatively, we could store two edges for each branch in the fragments, but I imaging that this gets weird when we draw the tree in the viewer.

File: src/skeleplex/graph/skeleton_graph.py (hunk 1 above)
Comment (kevinyamauchi) [reply src/skeleplex/graph/skeleton_graph.py:94]: Okay, thanks for looking. I think it's fine to leave as is.

File: src/skeleplex/graph/skeleton_graph.py (hunk 2 above)
Comment (kevinyamauchi) [reply src/skeleplex/graph/skeleton_graph.py:98]: Yeah, makes sense. Thanks

**Review by kevinyamauchi — APPROVED** [review]
LGTM! Thanks!
